FieldOpt/Optimization/optimizers/APPS.cpp

```diff
@@ -0,0 +1,126 @@
+/******************************************************************************
+   Created by einar on 11/21/16.
+   Copyright (C) 2016 Einar J.M. Baumann <einar.baumann@gmail.com>
+
+   This file is part of the FieldOpt project.
+
+   FieldOpt is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   FieldOpt is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with FieldOpt.  If not, see <http://www.gnu.org/licenses/>.
+******************************************************************************/
+
+#include <iostream>
+#include "APPS.h"
+#include "gss_patterns.hpp"
+#include "Utilities/stringhelpers.hpp"
+
+namespace Optimization {
+    namespace Optimizers {
+
+        APPS::APPS(Settings::Optimizer *settings, Case *base_case,
+                   Model::Properties::VariablePropertyContainer *variables, Reservoir::Grid::Grid *grid)
+                : GSS(settings, base_case, variables, grid) {
+            directions_ = GSSPatterns::Compass(num_vars_);
+            step_lengths_ = Eigen::VectorXd(directions_.size());
+            step_lengths_.fill(settings->parameters().initial_step_length);
+
+            assert(settings->parameters().max_queue_size >= 1.0);
+            max_queue_length_ = directions_.size() * settings->parameters().max_queue_size;
+            is_async_ = true;
+            iterate();
+        }
+
+        void APPS::iterate() {
+            if (inactive().size() > 0) {
+                case_handler_->AddNewCases(generate_trial_points(inactive()));
+                set_active(inactive());
+            }
+            iteration_++;
+            if (verbosity_level_ >= 1) print_state("ITERATION START");
+        }
+
+        void APPS::handleEvaluatedCase(Case *c) {
+            if (isImprovement(c)) successful_iteration(c);
+            else unsuccessful_iteration(c);
+        }
+
+        void APPS::successful_iteration(Case *c) {
+            tentative_best_case_ = c;
+            set_step_lengths(c->origin_step_length());
+            expand();
+            reset_active();
+            prune_queue();
+            if (verbosity_level_ >= 1) print_state("SUCCESSFUL ITERATION");
+            iterate();
+        }
+
+        void APPS::unsuccessful_iteration(Case *c) {
+            vector<int> unsuccessful_direction;
+            if (c->origin_case()->id() == tentative_best_case_->id()) {
```

Comment:
`tentative_best_case_` is also set  to the provided base case in the `Optimizer` constructor (which is called before the `APPS` and `GSS` constructors).